pkg/decoder/decoder.go

```diff
@@ -9,100 +9,128 @@ import (
+	for _, rule := range source.ProcessingRules {
+		switch rule.Type {
+		case config.MULTILINE:
+			delimiter = NewMultiLineDelimiter(outputChan, rule.Reg)
+		}
+	}
+	if delimiter == nil {
+		delimiter = NewSingleLineDelimiter(outputChan)
+	}
+
+	return New(inputChan, outputChan, delimiter)
 }
 
 // New returns an initialized Decoder
-func New(InputChan chan *Payload, OutputChan chan message.Message) *Decoder {
-	var msgBuf bytes.Buffer
+func New(InputChan chan *Input, OutputChan chan *Output, delimiter Delimiter) *Decoder {
+	var lineBuffer bytes.Buffer
 	return &Decoder{
 		InputChan:  InputChan,
 		OutputChan: OutputChan,
-		msgBuffer:  &msgBuf,
+		lineBuffer: &lineBuffer,
+		delimiter:  delimiter,
 	}
 }
 
 // Start starts the Decoder
 func (d *Decoder) Start() {
+	d.delimiter.Start()
 	go d.run()
 }
 
-// run lets the Decoder handle data coming from the InputChan
-func (d *Decoder) run() {
-	for data := range d.InputChan {
-		d.decodeIncomingData(data.content, data.offset)
-	}
-	d.OutputChan <- message.NewStopMessage()
-}
-
 // Stop stops the Decoder
 func (d *Decoder) Stop() {
 	close(d.InputChan)
+	d.delimiter.Stop()
 }
 
-var truncatedMsg = []byte("...TRUNCATED...")
-var truncatedLen = len(truncatedMsg)
-var maxMessageLen = config.MaxMessageLen - truncatedLen
-
-// sendBuffuredMessage flushes the buffer and sends the message
-func (d *Decoder) sendBuffuredMessage(offset int64) {
-	msg := make([]byte, d.msgBuffer.Len())
-	// d.msgBuffer.Bytes() returns a slice to the []byte, we thus need to copy it
-	copy(msg, d.msgBuffer.Bytes())
-	if len(msg) > 0 {
-		m := message.NewMessage(msg)
-		o := message.NewOrigin()
-		o.Offset = offset
-		m.SetOrigin(o)
-		d.OutputChan <- m
+// run lets the Decoder handle data coming from InputChan
+func (d *Decoder) run() {
+	for data := range d.InputChan {
+		d.decodeIncomingData(data.content)
 	}
-	d.msgBuffer.Reset()
+	d.OutputChan <- newOutputStop()
 }
 
-// decodeIncomingData splits raw data based on `\n`, creates and sends messages to a channel
-func (d *Decoder) decodeIncomingData(inBuf []byte, offset int64) {
-	var i, j = 0, 0
-	var maxj = maxMessageLen - d.msgBuffer.Len()
-	// Note: we will truncate messages of length MaxLen - truncatedLen
-	// instead of MaxLen. We'll live with it for now
-	for ; j < len(inBuf); j++ {
-		if inBuf[j] == '\n' {
-			d.msgBuffer.Write(inBuf[i:j])
-			d.sendBuffuredMessage(offset + int64(j+1))
-			i = j + 1 // +1 as we skip the `\n`
-			maxj = maxMessageLen - d.msgBuffer.Len()
-		} else if j == maxj {
-			d.msgBuffer.Write(inBuf[i:j])
-			d.msgBuffer.Write(truncatedMsg)
-			d.sendBuffuredMessage(offset + int64(j))
-			d.msgBuffer.Write(truncatedMsg)
+// decodeIncomingData splits raw data based on '\n', creates and processes new lines
+func (d *Decoder) decodeIncomingData(inBuf []byte) {
+	i, j := 0, 0
+	n := len(inBuf)
+	maxj := i + contentLenLimit - d.lineBuffer.Len()
+
+	for ; j < n; j++ {
+		if j == maxj {
+			// process the line as it is too long
```

Comment:
`process the line as it is too long` -> `don't wait for \n as the line is too long`?